### PR TITLE
imp: `CheckSubstituteAndUpdateState` for 08-wasm client

### PR DESCRIPTION
**Pull request by charleenfei**
## Description

closes: #4862 

*NOTE* the grandpa test will fail on this PR until contract changes are merged for the new msg type.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/light-clients/08-wasm/types/proposal_handle_test.go [hunk 1]
```diff
@@ -78,203 +78,6 @@ func (suite *TypesTestSuite) TestCheckSubstituteAndUpdateStateGrandpa() {
 	}
 }
 
-// func (suite *TypesTestSuite) TestCheckSubstituteAndUpdateStateBasicTendermint() {
```
Comment (charleenfei) [inline modules/light-clients/08-wasm/types/proposal_handle_test.go:81]: this test is covered in [this pr](https://github.com/cosmos/ibc-go/pull/4950/files)

File: modules/light-clients/08-wasm/types/proposal_handle.go [hunk 2]
```diff
@@ -10,34 +12,37 @@ import (
...
+	if !bytes.Equal(cs.CodeHash, substituteClientState.CodeHash) {
+		return errorsmod.Wrapf(types.ErrInvalidClient, "expected code hashes to be equal: expected %s, got %s", hex.EncodeToString(cs.CodeHash), hex.EncodeToString(substituteClientState.CodeHash))
+	}
+
 	store := newUpdateProposalWrappedStore(subjectClientStore, substituteClientStore, subjectPrefix, substitutePrefix)
 
+	// new payload message
 	payload := SudoMsg{
-		CheckSubstituteAndUpdateState: &CheckSubstituteAndUpdateStateMsg{},
+		MigrateClientStore: &MigrateClientStoreMsg{},
```
Comment (charleenfei) [inline modules/light-clients/08-wasm/types/proposal_handle.go:45]: see [this pr ](https://github.com/misko9/centauri/pull/5/files)for the contract changes pertaining to this new msg type

File: modules/light-clients/08-wasm/types/contract_api.go [hunk 3]
```diff
@@ -97,6 +98,9 @@ type VerifyUpgradeAndUpdateStateMsg struct {
 // CheckSubstituteAndUpdateStateMsg is a sudoMsg sent to the contract to check a given substitute client and update to its state.
 type CheckSubstituteAndUpdateStateMsg struct{}
```
Comment (colin-axner) [inline modules/light-clients/08-wasm/types/contract_api.go:99]: this should be removed?

File: modules/light-clients/08-wasm/types/proposal_handle.go [hunk 4]
```diff
@@ -13,31 +15,33 @@ import (
...
 		substitutePrefix = []byte("substitute/")
 	)
 
-	_, ok := substituteClient.(*ClientState)
+	substituteClientState, ok := substituteClient.(*ClientState)
 	if !ok {
 		return errorsmod.Wrapf(
 			clienttypes.ErrInvalidClient,
-			fmt.Sprintf("invalid substitute client state. expected type %T, got %T", &ClientState{}, substituteClient),
+			fmt.Sprintf("invalid substitute client state. Expected type %T, got %T", &ClientState{}, substituteClient),
```
Comment (colin-axner) [inline modules/light-clients/08-wasm/types/proposal_handle.go:30]:
```suggestion
			fmt.Sprintf("invalid substitute client state: expected type %T, got %T", &ClientState{}, substituteClient),
```

File: modules/light-clients/08-wasm/types/proposal_handle.go [hunk 5]
```diff
@@ -13,31 +15,33 @@ import (
...
 
+	// check that code hashes of subject client state and substitute client state match
+	// changing the code hash is only allowed through the migrate contract RPC endpoint
+	if !bytes.Equal(cs.CodeHash, substituteClientState.CodeHash) {
+		return errorsmod.Wrapf(clienttypes.ErrInvalidClient, "expected code hashes to be equal: expected %s, got %s", hex.EncodeToString(cs.CodeHash), hex.EncodeToString(substituteClientState.CodeHash))
+	}
+
 	store := newUpdateProposalWrappedStore(subjectClientStore, substituteClientStore, subjectPrefix, substitutePrefix)
 
+	// new payload message
```
Comment (colin-axner) [inline modules/light-clients/08-wasm/types/proposal_handle.go:42]: nit: not sure this comment provides much value

**Review by colin-axner — APPROVED** [review]

File: modules/light-clients/08-wasm/types/contract_api.go (hunk 3 above)
Comment (charleenfei) [reply modules/light-clients/08-wasm/types/contract_api.go:99]: good call, i will update this!

File: modules/light-clients/08-wasm/types/proposal_handle.go (hunk 4 above)
Comment (charleenfei) [reply modules/light-clients/08-wasm/types/proposal_handle.go:30]: fixed!

File: modules/light-clients/08-wasm/types/proposal_handle.go (hunk 5 above)
Comment (charleenfei) [reply modules/light-clients/08-wasm/types/proposal_handle.go:42]: will remove :)

File: modules/light-clients/08-wasm/types/proposal_handle.go (hunk 4 above)
Comment (DimitrisJim) [reply modules/light-clients/08-wasm/types/proposal_handle.go:30]: ah, I fixed the unecessary `Sprintf`, not the capitalization. happy merge-conflict Charly

Comment (charleenfei) [reply modules/light-clients/08-wasm/types/proposal_handle.go:30]: i think he just rm the `fmt.Sprintf` but i can add the colon 🤠

Comment (DimitrisJim) [reply modules/light-clients/08-wasm/types/proposal_handle.go]: team :fist_right: :fist_left: work

**Review by DimitrisJim — APPROVED** [review]

**Comment by damiannolan** [comment]
Looks like the mock vm needs to be updated with the type changes in order to build

**Pull request merged**